### ISSUE TITLE
Two questions concerning create_mask_rcnn_tf_record_multi.py 

### ISSUE THREAD

**Issue by captainst** [vijendra1125/Tensorfow_Object_detection_API-Custom_Mask_RCNN#13]
First, great job and congrats.

Question1: From the following code, I see that if a bounding box of one big object contains multiple other smaller objects of the same class (for example, if the objects are diagonal), then the mask of these smaller objects are included in the bounding box of this big object. I just want to make sure that it is an intended behavior.
code:
 ```
mask_np_black = mask_np*0
mask_np_black[int(ymin):int(ymax), int(xmin):int(xmax)] = mask_np[int(ymin):int(ymax), int(xmin):int(xmax)]
mask_remapped = (mask_np_black == mask_pixel[class_name]).astype(np.uint8)
masks.append(mask_remapped)
```

example:
![image](https://user-images.githubusercontent.com/23187023/59239115-09f85080-8c33-11e9-9749-d54ded8ffabf.png)

Question2: With the help of some 3rd party tools, for example, openCV, the create_mask_rcnn_tf_record.py could be modified without much difficulty to suit the case where multiple objects within the same class could be automatically handled without using LabelImg tool. If you want, I could make a PR on this.

Many thanks again.

**Comment by vijendraxsingh** [comment]
Hi, 

Thank you for pointing it out, mentioned behavior is not intended. It will great if you could fix this and make a PR, otherwise I will do it as soon as I get time. 

Thank you for your proposal to contribute. It will be great if you could work on removing the dependency on LabelImg tool. I also have an idea (not mature enough to cover all possible use cases though) to remove the dependency on LabelImg tool but I didn't get time to work with it. I would like to share my idea  with you (in case it could help you while you will be working with the idea you have). My idea is to use connected component algorithm for each class (or pixel color) in order to distinguish between different objects of same class (or pixel color) but drawback is the case where objects of same class have overlapping mask. It will be great if you could share your approach too and we discuss to finalize the approach.

**Comment by captainst** [comment]
Hi there,

Thank you for your quick reply. 

Concerning question 1, I have some doubt since actually I myself still do not quite understand the relationships between masks and bounding boxes. Do you mean that in the above example, the bigger bounding box should **only** contain the mask of the bigger rotated red rectangular, leaving out the smaller rotated rectangular?

Concerning question 2, I am thinking of using the thresholding and findcontour by openCV library. I will try to make a small test first to verify if this approach works. Then I'll make a PR for your verification. But the behavior of my first question will also affect the implementation.

**Comment by vijendraxsingh** [comment]
yes, the bigger bounding box should **only** contain the mask of the bigger rotated red rectangular, leaving out the smaller rotated rectangular because here we are drawing bounding box and mask for each individual object. 

Good Idea! Just find contour should also do, its just that it will also have same limitation as the one I mentioned in my last comment (fails in the case where we have overlapping mask). Unless until we find the solution for issue with overlapping masks of objects, it will not be good idea to take in the solution.

**Comment by captainst** [comment]
Thanks again.

By "overlapping mask", do you imply that **[A]** the same region of an image could belong to different objects? or, **[B]**, a situation shown in the following image that different objects "overlapped each other", but each pixel only belongs to one object:
![image](https://user-images.githubusercontent.com/23187023/59255807-35942e80-8c65-11e9-861d-4d656935515f.png)

Personally, I have not met situation [A], and I do not know if mask-rcnn support this case. But for situation [B], I think that openCV could handle it.

**Comment by vijendraxsingh** [comment]
Yes, I meant **[B]**. Looking forward to see the solution.

**Comment by captainst** [comment]
Hi Vijendra,

I have attached a mini-sample for your reference. 
I made myself a sample mask png with 3 circles: red, green, and blue, each overlapping with each other, like this:
![image](https://user-images.githubusercontent.com/23187023/59269196-186d5900-8c81-11e9-99e7-9bcfa6b09200.png)

This small sample will individual each partial body of the circle, and show all 3 partial bodies using cv2.imshow(). The rest of the work should be simple enough.

Please let me know if this is the behavior you were expecting. Cheers.
[overlapping_test.zip](https://github.com/vijendra1125/Custom-Mask-RCNN-using-Tensorfow-Object-detection-API/files/3276309/overlapping_test.zip)

**Comment by vijendraxsingh** [comment]
Hi, 

I think working with the case shown in the image below should be better because it also covers the case where the object of the same class is overlapping. 
![Screenshot 2019-06-11 18 52 06](https://user-images.githubusercontent.com/5885636/59275590-3d120280-8c7a-11e9-98ea-f8c0ad211843.png)
you have 2 red and 1 blue circle object in the above image. It covers both red to red overlap (overlap between same class) and red to blue overlap (overlap between different classes).

**Comment by captainst** [comment]
Hi Vijendra,

I think now I understand your concern. I have to admit that I haven't met this case before, though it comes now as an obvious question. 

I am not sure how exactly the mask-rcnn learns to segment these two red circles during training. Probably by using the bounding box data supplied with the mask? In this case, if we don't use the _LabelImage_, the machine has to learn to annotate the bounding box in autonomy. 

One way I would think of is to use watershed algorithm that comes freely by openCV library, assuming that the boundary between objects of the same class is clear, which could be a problem ...

Yet another approach I was thinking is to use the white color to annotate the border between objects of the same class, if they overlay each other. In this way, it could be easier for the machine to segment.

Yet another question: it is only possible to use rectangle as bounding box, we can not use a "bounding polygon" to work with mask-rcnn, right ?

Thanks again. 

Tom

**Comment by vijendraxsingh** [comment]
Hi Tom, 

One easy case which I could tell you is the balloon image you had attached above. There if we have balloon as a class then we will have to label all balloons as the same class and it will be overlapping. 

Good question regarding Mask-RCNN and because of how mask-RCNN architecture is built on top of Faster-RCNN, I agree with your opinion.
 
I think it's not safe to assume that the boundary between objects of the same class is clear.

Your other approach of annotating border will not be user-friendly but it will work well so I will say worth a shot. 

If we are using a pre-trained model given in Tensorflow API or any other pre-trained model which is trained with a rectangle as bounding-box then we could use rectangle only. If you want to use any other polygon then you will have to train your own model with the kind of polygon you want as a bounding box after doing some modifications in architecture.

**Comment by captainst** [comment]
Hi Vijendra,

Thank you for your explanation.

I came across [this ](https://engineering.matterport.com/splash-of-color-instance-segmentation-with-mask-r-cnn-and-tensorflow-7c761e238b46)article written by a matterport engineer which gives a good walk through over mask-rcnn.

This article did not mention the case you illustrated, but I am quite sure that his also used masks to generate bounding boxes for overlapped objects (balloons), as he wrote: 

> And, bounding boxes are generated dynamically from the masks.

Looking at his example of annotation, it seems that he used different colors to mask balloons that overlap each other:
[annotation image](https://cdn-images-1.medium.com/max/1600/1*OKE6wyZFfh2f_aZ3rd9BRw.png)

I will look into the dataset that he created to have a better understanding.

Tom

**Comment by captainst** [comment]
Hi Vijendra,

I looked into the balloon sample from matterport mask rcnn project. Practically it uses polygons to save the mask boundary for each object in the image. I think that's why there's no problem to generate bounding boxes in this case. Other than bounding boxes, the boundary of connected objects is already set by hand.

So in my personal opinion, since PixelAnnotationTool is great and ease to use and worth to stick to, it could be a work-around to ask user to manually add a boundary between the same objects overlapping with each other.

Tom

**Comment by vijendraxsingh** [comment]
Hi Tom, 

I agree that right now the only option we could see is to ask the user for drawing the boundary, it will be great if you could implement this in a separate file and make a PR. I will review it.

**Comment by captainst** [comment]
Hi Vijendra,

I have tried the "drawing boundary manully" approach and it seems that PixelAnnotationTool lacks some native support for it. See question I raised _https://github.com/abreheret/PixelAnnotationTool/issues/39_ 

So I have been re-thinking this issue. there might be another approach, to better make use of the capability of PixelAnnotationTool.

For the overlapped objects, no matter if they belong to the same type or no, use different color to mark them. And in your **create_mask_rcnn_tf_record.py** file, simply modify your color dictionary like this:

`mask_pixel = {'redcircle':[76,149], 'bluecircle':[29]}`

So instead of assigning one color to one object, use an array with multiple colors for one object. In the above example, I use 76 and 149 to define "redcircle" and 29 for "bluecircle". In this case, as long as the color assignments are consistant throughout the whole dataset, it would be fine.

And it requires some minimum modifications to the original code. What do you think ?

**Comment by vijendraxsingh** [comment]
Hi Tom, 
Sorry for the late reply because of unavailability. I agree with your idea,  keeping different color for all overlapping objects will be good in order to avoid the need of boundary whereas the choice of using different colors for all or keeping enough different color to avoid overlapping objects of same class having the same color could be left on the user (same code will work for either choice made by user). I reviewed the PR you made, looks good.
